internal/state/storage/storage.go

```diff
@@ -0,0 +1,392 @@
+		for {
+			numRead, err := reader.Read(buffer[index:])
+			size += numRead
+
+			if err != nil {
+				if err == io.EOF { //nolint:errorlint
+					return nil, false, nil
+				}
+
+				return nil, false, fmt.Errorf("failed to read bytes: %w", err)
+			}
+
+			clusterSnapshot, offset, err := decodeClusterSnapshot(buffer)
+			if err != nil {
+				return nil, false, fmt.Errorf("failed to decode cluster: %w", err)
+			}
+
+			clusterEndIndex := len(buffer) + offset
+
+			if clusterSnapshot == nil { // we need to read more data
+				buffer = slices.Grow(buffer, offset)
+				buffer = buffer[:clusterEndIndex]
+				index += numRead
+
+				continue
+			}
+
+			// prepare the buffer for the next iteration
+			buffer = slices.Delete(buffer, 0, clusterEndIndex) // delete the bytes which belong to the processed cluster
+			index = len(buffer)                                // set the start index for the next iteration - here, the remaining bytes in the buffer belong to the next cluster
+			buffer = buffer[:bufferSize]                       // reset the buffer to the initial size
+
+			// update stats
+			numClusters++
+			numAffiliates += len(clusterSnapshot.Affiliates)
+
+			for _, affiliate := range clusterSnapshot.Affiliates {
+				numEndpoints += len(affiliate.Endpoints)
+			}
+
+			return clusterSnapshot, true, nil
+		}
+	}); err != nil {
+		return SnapshotStats{}, fmt.Errorf("failed to import clusters: %w", err)
+	}
+
+	return SnapshotStats{
+		Size:          size,
+		NumClusters:   numClusters,
+		NumAffiliates: numAffiliates,
+		NumEndpoints:  numEndpoints,
+	}, nil
+}
+
+// Export exports all clusters' states into the given writer.
+func (storage *Storage) Export(writer io.Writer) (SnapshotStats, error) {
+	numClusters := 0
+	numAffiliates := 0
+	numEndpoints := 0
+	size := 0
+
+	var buffer []byte
+
+	// marshal the clusters in a streaming manner and export them into the writer
+	if err := storage.state.ExportClusterSnapshots(func(snapshot *storagepb.ClusterSnapshot) error {
+		var err error
+
+		buffer, err = encodeClusterSnapshot(buffer, snapshot)
+		if err != nil {
+			return fmt.Errorf("failed to encode cluster: %w", err)
+		}
+
+		written, err := writer.Write(buffer)
+		if err != nil {
+			return fmt.Errorf("failed to write cluster: %w", err)
+		}
+
+		// prepare the buffer for the next iteration - reset it
+		buffer = buffer[:0]
+
+		// update stats
+		size += written
+		numClusters++
+		numAffiliates += len(snapshot.Affiliates)
+
+		for _, affiliate := range snapshot.Affiliates {
+			numEndpoints += len(affiliate.Endpoints)
+		}
+
+		return nil
+	}); err != nil {
+		return SnapshotStats{}, fmt.Errorf("failed to snapshot clusters: %w", err)
+	}
+
+	return SnapshotStats{
+		Size:          size,
+		NumClusters:   numClusters,
+		NumAffiliates: numAffiliates,
+		NumEndpoints:  numEndpoints,
+	}, nil
+}
+
+func getTempFile(dst string) (*os.File, error) {
+	tmpFile, err := os.OpenFile(dst+".tmp", os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o666)
```

Comment:
if we use `O_EXCL`, we can easily break things if there's a `.tmp` file already in the filesystem (e.g. if crashed).

I would rather use `O_TRUNC`.